net: config: sntp: Asynchronous resynchronization

rlubos opened the pull request:
Make SNTP resynchronization asynchronous, so that it doesn't block the system work queue while waiting for response.

Fixes #99634

rlubos committed:
net: config: sntp: Asynchronous resynchronization

Make SNTP resynchronization asynchronous, so that it doesn't block the
system work queue while waiting for response.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

jeppenodgaard commented on subsys/net/lib/config/init_clock_sntp.c:
```diff
@@ -112,10 +112,230 @@ int net_init_clock_via_sntp(void)
+out:
+	k_work_reschedule(
+		&sntp_resync_work_handle,
+		(ret < 0) ? K_SECONDS(CONFIG_NET_CONFIG_SNTP_INIT_RESYNC_ON_FAILURE_INTERVAL)
```
```suggestion
		(ret < 0) ? RESYNC_FAILED_INTERVAL
```

jeppenodgaard commented on subsys/net/lib/config/init_clock_sntp.c:
```diff
@@ -112,10 +112,230 @@ int net_init_clock_via_sntp(void)
+
+#define RESYNC_FAILED_INTERVAL K_SECONDS(CONFIG_NET_CONFIG_SNTP_INIT_RESYNC_ON_FAILURE_INTERVAL)
+
+static void sntp_async_timeout(struct k_work *work)
```
Nit: AFAIK you can use `__unused` instead of `ARG_UNUSED()`
```suggestion
static void sntp_async_timeout(struct k_work __unused *work)
```

rlubos replied:
I didn't see a single instance in our codebase where `__unused` was used with function argument, I'll stick to `ARG_UNUSED()` which is a common pattern.

jeppenodgaard commented on subsys/net/lib/config/init_clock_sntp.c:
```diff
@@ -112,10 +112,230 @@ int net_init_clock_via_sntp(void)
+
+	sntp_set_rtc(&tspec);
+
+	LOG_DBG("Time synced using SNTP, SNTP Time: %llu", ts.seconds);
```
```suggestion
	LOG_DBG("Time synced using SNTP, SNTP Time: %" PRIu64, ts.seconds);
```

rlubos replied:
Fixed in helper function

jeppenodgaard commented on subsys/net/lib/config/init_clock_sntp.c:
```diff
@@ -112,10 +112,230 @@ int net_init_clock_via_sntp(void)
+	ret = dns_get_addr_info(CONFIG_NET_CONFIG_SNTP_INIT_SERVER,
+				type, NULL, dns_result_cb,
+				NULL, CONFIG_NET_CONFIG_SNTP_INIT_TIMEOUT);
+	if (ret < 0) {
```
```suggestion
	return ret;
```
and remove the lines below.

rlubos replied:
This code was refactored anyway due to no DNS_RESOLVER fallback.

maass-hamburg commented on subsys/net/lib/config/init_clock_sntp.c:
```diff
@@ -112,10 +112,230 @@ int net_init_clock_via_sntp(void)
+	k_work_reschedule(
+		&sntp_resync_work_handle,
+		(ret < 0) ? K_SECONDS(CONFIG_NET_CONFIG_SNTP_INIT_RESYNC_ON_FAILURE_INTERVAL)
+			  : K_SECONDS(CONFIG_NET_CONFIG_SNTP_INIT_RESYNC_INTERVAL));
```
The regular `net_init_clock_via_sntp` is no longer used, when CONFIG_NET_CONFIG_SNTP_INIT_USE_CONNECTION_MANAGER is enabled, as it uses the resync variant always.

maass-hamburg commented on subsys/net/lib/config/init_clock_sntp.c:
```diff
@@ -112,10 +112,230 @@ int net_init_clock_via_sntp(void)
+		type = DNS_QUERY_TYPE_AAAA;
+	}
+
+	ret = dns_get_addr_info(CONFIG_NET_CONFIG_SNTP_INIT_SERVER,
```
This also works when DNS is enabled, but the Kconfig is a IP address right?
Just like `getaddrinfo()`

jeppenodgaard commented on subsys/net/lib/config/init_clock_sntp.c:
```diff
@@ -112,10 +112,230 @@ int net_init_clock_via_sntp(void)
+		goto out;
+	}
+
+	tspec.tv_sec = ts.seconds;
```
These lines are very identical to `net_init_clock_via_sntp`. Can you wrap them in a function?

rlubos replied:
Added helper function for setting clocs.

jeppenodgaard reviewed (commented):
LGTM. Non-blocking comments.
I will try it out now.

rlubos committed:
net: config: sntp: Add helper function to set clocks

Move the code responsible for setting clocks into a helper function,
as it'll be reused with asynchronous resynchronization.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: config: sntp: Add helper symbols for resync timeouts

For more concise code.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

jeppenodgaard reviewed (dismissed):
To compile `DNS_RESOLVER` must be enabled. This was not required before.

Oh - and thank you for opening this PR so quickly. I like this solution :)

rlubos commented:
Beside some smaller changes addressing review comments I've added a fallback for a scenario where DNS_RESOLVED is not enabled in the system, please re-review.

jeppenodgaard reviewed (commented):
LGTM!

L4 connect -> disconnect -> connect will always result in a SNTP query:
```bash
[00:01:00.715,000] <dbg> net_config: sntp_set_clocks: Time synced using SNTP, SNTP Time: 1763728695
[00:01:02.364,000] <inf> phy_mii: PHY (0) is down
[00:01:03.965,000] <inf> phy_mii: PHY (0) Link speed 100 Mb, full duplex
...
[00:01:04.812,000] <dbg> net_config: sntp_set_clocks: Time synced using SNTP, SNTP Time: 1763728699
```
I am pretty sure this has always been an issue due to:
```c
	if (mgmt_event == NET_EVENT_L4_CONNECTED) {
		k_work_reschedule(&sntp_resync_work_handle, K_NO_WAIT);
	} else if (mgmt_event == NET_EVENT_L4_DISCONNECTED) {
		k_work_cancel_delayable(&sntp_resync_work_handle);
	}
```

It is probably too corner case to report/fix.

rlubos commented:
> I am pretty sure this has always been an issue due to:
> 
> ```c
> 	if (mgmt_event == NET_EVENT_L4_CONNECTED) {
> 		k_work_reschedule(&sntp_resync_work_handle, K_NO_WAIT);
> 	} else if (mgmt_event == NET_EVENT_L4_DISCONNECTED) {
> 		k_work_cancel_delayable(&sntp_resync_work_handle);
> 	}
> ```
> 
> It is probably too corner case to report/fix.

Is that really an issue though? What's the problem with the query being sent when the interface goes up?

Anyway, it has nothing to do with this PR, please note there's still a NACK from your side.

maass-hamburg commented on subsys/net/lib/config/init_clock_sntp.c:
```diff
@@ -82,40 +84,273 @@ static void sntp_set_rtc(__maybe_unused const struct timespec *tspec)
+		sntp_addr.ss_family = NET_AF_INET;
+		net_ipv4_addr_copy_raw(net_sin(net_sad(&sntp_addr))->sin_addr.s4_addr,
+					       iface->config.dhcpv4.ntp_addr.s4_addr);
+		net_sin(net_sad(&sntp_addr))->sin_port = net_htons(123);
```
'123' should be a define and it should maybe be said where this value comes from

rlubos commented on subsys/net/lib/config/init_clock_sntp.c:
```diff
@@ -82,40 +84,273 @@ static void sntp_set_rtc(__maybe_unused const struct timespec *tspec)
+					NULL, CONFIG_NET_CONFIG_SNTP_INIT_TIMEOUT);
+		if (ret < 0) {
+			LOG_ERR("Failed to initiate DNS query for SNTP server (%d)", ret);
+		}
```
Done

jeppenodgaard commented on subsys/net/lib/config/init_clock_sntp.c:
```diff
@@ -112,10 +112,230 @@ int net_init_clock_via_sntp(void)
+	}
+
+	k_work_reschedule(&sntp_async_timeout_work,
+			  K_MSEC(CONFIG_NET_CONFIG_SNTP_INIT_TIMEOUT));
```
> Not sure I follow the exact issue here, even if for whatever reason the packet was sent, what's the problem?

The packet would default be sent again after 3 s which is not allowed:
https://github.com/zephyrproject-rtos/zephyr/pull/99754#issuecomment-3569583001

Resolving as it should never be the case.

jeppenodgaard commented on subsys/net/lib/config/init_clock_sntp.c:
```diff
@@ -82,40 +84,276 @@ static void sntp_set_rtc(__maybe_unused const struct timespec *tspec)
+		if (sntp_addr.ss_family == NET_AF_INET) {
+			sntp_addrlen = sizeof(struct net_sockaddr_in);
+			net_sin(net_sad(&sntp_addr))->sin_port = net_htons(SNTP_SERVER_PORT);
+		} else {
```
Not sure why the build command does not trigger another warning.

I need to change this as well with my custom setup:
```diff
diff --git a/subsys/net/lib/config/init_clock_sntp.c b/subsys/net/lib/config/init_clock_sntp.c
index d8e259718a6..895b4fdebab 100644
--- a/subsys/net/lib/config/init_clock_sntp.c
+++ b/subsys/net/lib/config/init_clock_sntp.c
@@ -232,36 +232,23 @@ static void dns_result_cb(enum dns_resolve_status status,
        }

        if (status == DNS_EAI_INPROGRESS && info != NULL) {
-               switch (info->ai_family) {
-               case NET_AF_INET:
-                       if (sntp_addrlen > 0) {
-                               /* Already got address, skip others. */
-                               break;
-                       }
+               if (sntp_addrlen > 0) {
+                       /* Already got address, skip others. */
+                       return;
+               }

+               if (IS_ENABLED(CONFIG_NET_IPV4) && info->ai_family == NET_AF_INET) {
                        sntp_addrlen = info->ai_addrlen;
                        sntp_addr.ss_family = NET_AF_INET;
                        net_ipv4_addr_copy_raw(net_sin(net_sad(&sntp_addr))->sin_addr.s4_addr,
                                               net_sin(&info->ai_addr)->sin_addr.s4_addr);
                        net_sin(net_sad(&sntp_addr))->sin_port = net_htons(123);
-                       break;
-
-               case NET_AF_INET6:
-                       if (sntp_addrlen > 0) {
-                               /* Already got address, skip others. */
-                               break;
-                       }
-
-                       sntp_addrlen = info->ai_addrlen;
+               } else if (IS_ENABLED(CONFIG_NET_IPV6) && info->ai_family == NET_AF_INET6) {
+                                               sntp_addrlen = info->ai_addrlen;
                        sntp_addr.ss_family = NET_AF_INET6;
                        net_ipv6_addr_copy_raw(net_sin6(net_sad(&sntp_addr))->sin6_addr.s6_addr,
                                               net_sin6(&info->ai_addr)->sin6_addr.s6_addr);
                        net_sin6(net_sad(&sntp_addr))->sin6_port = net_htons(123);
-                       break;
-
-               default:
-                       /* Ignore other records. */
-                       break;
                }
        }
 }
```

jeppenodgaard replied:
https://github.com/zephyrproject-rtos/zephyr/pull/99754#discussion_r2555511680 still needs to be addressed.

maass-hamburg commented on subsys/net/lib/config/init_clock_sntp.c:
```diff
@@ -82,40 +84,263 @@ static void sntp_set_rtc(__maybe_unused const struct timespec *tspec)
+		} else if (IS_ENABLED(CONFIG_NET_IPV6)) {
+			sntp_addrlen = sizeof(struct net_sockaddr_in6);
+			net_sin6(net_sad(&sntp_addr))->sin6_port = net_htons(SNTP_SERVER_PORT);
+		}
```
why not using ``net_port_set_default()`` like in ``sntp_simple_addr()``

rlubos replied:
Totally forgot we had function like this. Good point about not overwriting custom port number, I've updated both cases to use `net_port_set_default()` instead of setting port manually.
